# Visualización de datos

## Introducción a la visualización de datos con Python

- La visualización de datos es el arte y la ciencia de contar historias cautivadoras con datos.

- <b>¿Por qué Python?</b>
    
- <b>Python</b> realiza cálculos numéricos y científicos avanzados con librerías como: <b>numpy</b> y <b>scipy</b>, alberga una amplia gama de métodos de aprendizaje automático gracias a la disponibilidad del paquete <b>scikit-learn</b>

- Proporciona una gran interfaz para manipulación de **big data** gracias a la disponibilidad del paquete `pandas` y su compatibilidad con `Apache Spark`

- Genera gráficos y figuras estéticamente correctas con librerías como <b>seaborn, plotly</b>, etc.

### Manejo de datos con pandas DataFrame

- La biblioteca **pandas** es un conjunto de herramientas de código abierto extremadamente ingenioso para manejar manipular y analizar datos estructurados. Las tablas de datos se pueden almacenar en el objeto **DataFrame** disponible en **pandas**, y los datos en múltiples formatos (por ejemplo, **.csv, .tsv, .xlsx** y **.json**) pueden leerse directamente en un **DataFrame**.  

### Ejercicio 1: Lectura de datos desde archivos

- En este ejercicio, leeremos de un conjunto de datos `diamonds.csv`
- Abre un `notebook` **jupyter** y carga la librería **pandas**

In [1]:
import pandas as pd

In [2]:
import warnings
warnings.filterwarnings('ignore')

- Especifique la **UR**L del conjunto de datos:

In [3]:
diamonds_url = "https://raw.githubusercontent.com/lihkir/Uninorte/main/AppliedStatisticMS/DataVisualizationRPython/Lectures/Python/PythonDataSets/diamonds.csv"

- Leer los archivos de la **URL** en el **DataFrame** de **pandas**

In [4]:
diamonds_df = pd.read_csv(diamonds_url)

In [5]:
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

- Las columnas del `dataframe` correponden a: 

    - `price`: Precio en dólares estadounidenses (\$326--\$18.823)
    - `carat`: Peso en quilates del diamante (0,2--5,01)
    - `cut`: Calidad de la talla (regular, buena, muy buena, superior, ideal)
    - `color`: Color del diamante, de J (peor) a D (mejor)
    - `clarity`: claridad una medida de la claridad del diamante (I1 (peor), SI2, SI1, VS2, VS1, VVS2, VVS1, IF (mejor))
    - `x`: Longitud x en mm (0--10,74)
    - `y`: Anchura en mm (0--58,9)
    - `z`: Profundidad en mm (0--31,8)
    - `depth`: Profundidad total porcentaje = z / media(x, y) = 2 * z / (x + y) (43--79)
    - `table`: Tabla de la anchura de la parte superior del diamante en relación con el punto más ancho (43--95)

- Utilice el parámetro **usecols** si sólo necesita leer columnas específicas

In [6]:
diamonds_df_specific_cols = pd.read_csv(diamonds_url, usecols=['carat','cut','color','clarity'])

In [7]:
diamonds_df_specific_cols.head()

carat      cut color clarity
0   0.23    Ideal     E     SI2
1   0.21  Premium     E     SI1
2   0.23     Good     E     VS1
3   0.29  Premium     I     VS2
4   0.31     Good     J     SI2

### Ejercicio 2: Observación y descripción de datos

- En este ejercicio, veremos cómo observar y describir datos en un **DataFrame**. Volveremos a utilizar el conjunto de datos de diamantes
- Cargue la librería pandas:

In [8]:
import pandas as pd

- Leer los archivos de la **URL** en el **DataFrame** de **pandas**

In [9]:
diamonds_df = pd.read_csv(diamonds_url)

- Observe los datos utilizando la función **head**:

In [10]:
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

Los datos contienen diferentes características de los diamantes, como **quilates, calidad de corte, color** y **precio**: `carat, cut, color, clarity, depth, table, price`  como columnas. Ahora, **corte, claridad** y **color** son **variables categóricas**, y **x, y, z, profundidad, tabla** y **precio** son **variables continuas**. Mientras que las variables categóricas toman como valores categorías/nombres únicos, los valores continuos toman números reales como valores.

- Contar el número de filas y columnas en el **DataFrame** utilizando la función **shape** 

In [11]:
diamonds_df.shape

(53940, 10)

- Resumir las columnas utilizando `describe()` para obtener la distribución de las variables, incluyendo la media, la mediana, el mínimo, el máximo y los diferentes cuartiles

In [12]:
diamonds_df.describe()

carat         depth         table         price             x  \
count  53940.000000  53940.000000  53940.000000  53940.000000  53940.000000   
mean       0.797940     61.749405     57.457184   3932.799722      5.731157   
std        0.474011      1.432621      2.234491   3989.439738      1.121761   
min        0.200000     43.000000     43.000000    326.000000      0.000000   
25%        0.400000     61.000000     56.000000    950.000000      4.710000   
50%        0.700000     61.800000     57.000000   2401.000000      5.700000   
75%        1.040000     62.500000     59.000000   5324.250000      6.540000   
max        5.010000     79.000000     95.000000  18823.000000     10.740000   

                  y             z  
count  53940.000000  53940.000000  
mean       5.734526      3.538734  
std        1.142135      0.705699  
min        0.000000      0.000000  
25%        4.720000      2.910000  
50%        5.710000      3.530000  
75%        6.540000      4.040000  
max       58.900000     31.800000

```{figure} ./imgs/sesgo_media_mediana.png
:name: sesgo_media_mediana
:align: center

Represestanción del sesgo dependiendo de las medidas de tendencia central.
```

- Esto funciona para las variables continuas. Sin embargo, para las variables categóricas, necesitamos utilizar el parámetro **include=object**. 

In [13]:
diamonds_df.describe(include=object)

cut  color clarity
count   53940  53940   53940
unique      5      7       8
top     Ideal      G     SI1
freq    21551  11292   13065

- Para obtener información sobre el conjunto de datos, utilice el método `info()`:

In [14]:
diamonds_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   carat    53940 non-null  float64
 1   cut      53940 non-null  object 
 2   color    53940 non-null  object 
 3   clarity  53940 non-null  object 
 4   depth    53940 non-null  float64
 5   table    53940 non-null  float64
 6   price    53940 non-null  int64  
 7   x        53940 non-null  float64
 8   y        53940 non-null  float64
 9   z        53940 non-null  float64
dtypes: float64(6), int64(1), object(3)
memory usage: 4.1+ MB


- Podemos acceder a la columna corte del **DataFrame diamonds_df** con **diamonds_df.cut** o **diamonds_df['cut']**
- Ahora, ¿qué tal si seleccionamos todas las filas correspondientes a los diamantes que tienen la talla **Ideal** y almacenarlas en un **DataFrame** separado? Podemos seleccionarlas utilizando la función **loc** para seleccionarlos:

In [15]:
diamonds_low_df = diamonds_df.loc[diamonds_df['cut']=='Ideal']
diamonds_low_df.head()

carat    cut color clarity  depth  table  price     x     y     z
0    0.23  Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
11   0.23  Ideal     J     VS1   62.8   56.0    340  3.93  3.90  2.46
13   0.31  Ideal     J     SI2   62.2   54.0    344  4.35  4.37  2.71
16   0.30  Ideal     I     SI2   62.0   54.0    348  4.31  4.34  2.68
39   0.33  Ideal     I     SI2   61.8   55.0    403  4.49  4.51  2.78

### Ejercicio 3: Añadir nuevas columnas al DataFrame

- En este ejercicio, vamos a añadir nuevas columnas al conjunto de datos de diamantes en la biblioteca **pandas**. Empezaremos con la adición simple de columnas y luego avanzaremos y veremos la adición condicional de columnas. Para ello, vamos a seguir los siguientes pasos:

- Cargue la biblioteca `pandas`

In [16]:
import pandas as pd

- Leer los archivos de la **URL** en el **DataFrame** de **pandas**

In [17]:
diamonds_df = pd.read_csv(diamonds_url)
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

- Añade una columna `price_per_carat` al `DataFrame`. En este ejemplo el precio por quilates. Del mismo modo, también podemos utilizar la suma, la resta y otros operadores matemáticos sobre dos columnas numéricas.

In [18]:
diamonds_df['price_per_carat'] = diamonds_df['price']/diamonds_df['carat']

- Llame a la función **head** de **DataFrame** para comprobar si la nueva columna se ha añadido como como se esperaba:

In [19]:
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z  \
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43   
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31   
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31   
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63   
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75   

   price_per_carat  
0      1417.391304  
1      1552.380952  
2      1421.739130  
3      1151.724138  
4      1080.645161

- Ahora, veremos la adición condicional de columnas. Vamos a intentar añadir una columna basada en el valor de `price_per_carat`, digamos que todo lo que sea más de 3500 como alto (codificado como 1) y todo lo que sea inferior a 3500 como bajo (codificado como 0).

In [20]:
import numpy as np

diamonds_df['price_per_carat_is_high'] = np.where(diamonds_df['price_per_carat'] > 3500, 1, 0)
diamonds_df.tail()

carat        cut color clarity  depth  table  price     x     y     z  \
53935   0.72      Ideal     D     SI1   60.8   57.0   2757  5.75  5.76  3.50   
53936   0.72       Good     D     SI1   63.1   55.0   2757  5.69  5.75  3.61   
53937   0.70  Very Good     D     SI1   62.8   60.0   2757  5.66  5.68  3.56   
53938   0.86    Premium     H     SI2   61.0   58.0   2757  6.15  6.12  3.74   
53939   0.75      Ideal     D     SI2   62.2   55.0   2757  5.83  5.87  3.64   

       price_per_carat  price_per_carat_is_high  
53935      3829.166667                        1  
53936      3829.166667                        1  
53937      3938.571429                        1  
53938      3205.813953                        0  
53939      3676.000000                        1

### Ejercicio 4: Aplicación de funciones a las columnas de DataFrame

- En este ejercicio, `consideraremos un escenario en el que el precio de los diamantes ha aumentado y queremos aplicar un factor de incremento de 1.3 al precio de todos los diamantes en nuestro registro (aumento del 30%)`. Podemos conseguirlo aplicando una sencilla función. A continuación, redondearemos el precio de los diamantes hasta su tope. Para ello, vamos a seguir los siguientes pasos

- Cargue la biblioteca `pandas`

In [21]:
import pandas as pd

- Leer los archivos de la **URL** en el **DataFrame** de **pandas**

In [22]:
diamonds_df = pd.read_csv(diamonds_url)
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

- Aplique una función simple en las columnas utilizando el siguiente código:

In [23]:
diamonds_df['price'] = diamonds_df['price']*1.3

In [24]:
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0  423.8  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0  423.8  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0  425.1  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0  434.2  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0  435.5  4.34  4.35  2.75

- Aplicar la función `math.ceil` para `redondear el precio de los diamantes hacia arriba`

In [25]:
import math

diamonds_df['rounded_price'] = diamonds_df['price'].apply(math.ceil)
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z  \
0   0.23    Ideal     E     SI2   61.5   55.0  423.8  3.95  3.98  2.43   
1   0.21  Premium     E     SI1   59.8   61.0  423.8  3.89  3.84  2.31   
2   0.23     Good     E     VS1   56.9   65.0  425.1  4.05  4.07  2.31   
3   0.29  Premium     I     VS2   62.4   58.0  434.2  4.20  4.23  2.63   
4   0.31     Good     J     SI2   63.3   58.0  435.5  4.34  4.35  2.75   

   rounded_price  
0            424  
1            424  
2            426  
3            435  
4            436

- Puede haber ocasiones en las que tenga que escribir su propia función para realizar la tarea que desea llevar a cabo. 
- Por ejemplo, digamos que quiere añadir otra columna al **DataFrame** indicando el precio redondeado de los diamantes al múltiplo de 100 (igual o superior al precio).
- Utilice la función `lambda` como sigue para redondear el precio de los diamantes al múltiplo de 100 más cercano

In [26]:
import math

diamonds_df['rounded_price_to_100multiple'] = diamonds_df['price'].apply(lambda x: 100*math.ceil(x/100))
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z  \
0   0.23    Ideal     E     SI2   61.5   55.0  423.8  3.95  3.98  2.43   
1   0.21  Premium     E     SI1   59.8   61.0  423.8  3.89  3.84  2.31   
2   0.23     Good     E     VS1   56.9   65.0  425.1  4.05  4.07  2.31   
3   0.29  Premium     I     VS2   62.4   58.0  434.2  4.20  4.23  2.63   
4   0.31     Good     J     SI2   63.3   58.0  435.5  4.34  4.35  2.75   

   rounded_price  rounded_price_to_100multiple  
0            424                           500  
1            424                           500  
2            426                           500  
3            435                           500  
4            436                           500

- Por supuesto, no todas las funciones pueden ser escritas en una sola línea y es importante saber cómo incluir funciones definidas por el usuario en la función `apply`. Vamos a escribir el mismo código con una función definida por el usuario para ilustrarlo.

In [27]:
import math

def get_100_multiple_ceil(x):
    y = 100*math.ceil(x/100)
    return y
 
diamonds_df['rounded_price_to_100multiple']=diamonds_df['price'].apply(get_100_multiple_ceil)
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z  \
0   0.23    Ideal     E     SI2   61.5   55.0  423.8  3.95  3.98  2.43   
1   0.21  Premium     E     SI1   59.8   61.0  423.8  3.89  3.84  2.31   
2   0.23     Good     E     VS1   56.9   65.0  425.1  4.05  4.07  2.31   
3   0.29  Premium     I     VS2   62.4   58.0  434.2  4.20  4.23  2.63   
4   0.31     Good     J     SI2   63.3   58.0  435.5  4.34  4.35  2.75   

   rounded_price  rounded_price_to_100multiple  
0            424                           500  
1            424                           500  
2            426                           500  
3            435                           500  
4            436                           500

### Ejercicio 5: Aplicación de funciones en varias columnas

- Supongamos que estamos interesados en comprar diamantes que tengan **Ideal** **cut** y **color** **D** (totalmente incoloro). Este ejercicio consiste en añadir una nueva columna, **desired**, al **DataFrame**, cuyo valor será **yes** si se cumplen nuestros criterios y **no** si no se cumplen. Veamos cómo lo hacemos:

- Cargue la biblioteca **pandas**

In [28]:
import pandas as pd

- Leer los archivos de la **URL** en el **DataFrame** de **pandas**

In [29]:
diamonds_df_exercise = pd.read_csv(diamonds_url)
diamonds_df_exercise.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

- Write a function to determine whether a record, **x**, is desired or not:

In [30]:
def is_desired(x):
    bool_var = 'yes' if (x['cut']=='Ideal' and x['color']=='D') else 'no'
    return bool_var

- Utilice la función `apply` para añadir la nueva columna, **desired**:

In [31]:
diamonds_df_exercise['desired'] = diamonds_df_exercise.apply(is_desired, axis = 'columns')
diamonds_df_exercise.tail()

carat        cut color clarity  depth  table  price     x     y     z  \
53935   0.72      Ideal     D     SI1   60.8   57.0   2757  5.75  5.76  3.50   
53936   0.72       Good     D     SI1   63.1   55.0   2757  5.69  5.75  3.61   
53937   0.70  Very Good     D     SI1   62.8   60.0   2757  5.66  5.68  3.56   
53938   0.86    Premium     H     SI2   61.0   58.0   2757  6.15  6.12  3.74   
53939   0.75      Ideal     D     SI2   62.2   55.0   2757  5.83  5.87  3.64   

      desired  
53935     yes  
53936      no  
53937      no  
53938      no  
53939     yes

### Ejercicio 6: Eliminación de columnas de un DataFrame 

- Por último, vamos a ver cómo eliminar columnas de un `DataFrame` de `pandas`. Por ejemplo, borraremos las columnas `rounded_price` y `rounded_price_to_100multiple`

- Cargue la biblioteca `pandas`

In [32]:
import pandas as pd

- Leer los archivos de la `URL` en el `DataFrame` de `pandas`

In [33]:
diamonds_df = pd.read_csv(diamonds_url)
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

- Añadir la columna `price_per_carat` al `DataFrame`

In [34]:
diamonds_df['price_per_carat'] = diamonds_df['price']/diamonds_df['carat']
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z  \
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43   
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31   
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31   
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63   
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75   

   price_per_carat  
0      1417.391304  
1      1552.380952  
2      1421.739130  
3      1151.724138  
4      1080.645161

- Utilice la función `np.where` del paquete `numpy` de Python:

In [35]:
import numpy as np

diamonds_df['price_per_carat_is_high'] = np.where(diamonds_df['price_per_carat'] > 3500, 1, 0)
diamonds_df.tail()

carat        cut color clarity  depth  table  price     x     y     z  \
53935   0.72      Ideal     D     SI1   60.8   57.0   2757  5.75  5.76  3.50   
53936   0.72       Good     D     SI1   63.1   55.0   2757  5.69  5.75  3.61   
53937   0.70  Very Good     D     SI1   62.8   60.0   2757  5.66  5.68  3.56   
53938   0.86    Premium     H     SI2   61.0   58.0   2757  6.15  6.12  3.74   
53939   0.75      Ideal     D     SI2   62.2   55.0   2757  5.83  5.87  3.64   

       price_per_carat  price_per_carat_is_high  
53935      3829.166667                        1  
53936      3829.166667                        1  
53937      3938.571429                        1  
53938      3205.813953                        0  
53939      3676.000000                        1

- Aplicar una función compleja para `redondear el precio de los diamantes hacia arriba`

In [36]:
import math

diamonds_df['rounded_price'] = diamonds_df['price'].apply(math.ceil)
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z  \
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43   
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31   
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31   
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63   
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75   

   price_per_carat  price_per_carat_is_high  rounded_price  
0      1417.391304                        0            326  
1      1552.380952                        0            326  
2      1421.739130                        0            327  
3      1151.724138                        0            334  
4      1080.645161                        0            335

- Escribe un código para crear una función definida por el usuario:

In [37]:
import math

def get_100_multiple_ceil(x):
    y = math.ceil(x/100)*100
    return y
 
diamonds_df['rounded_price_to_100multiple']=diamonds_df['price'].apply(get_100_multiple_ceil)
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z  \
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43   
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31   
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31   
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63   
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75   

   price_per_carat  price_per_carat_is_high  rounded_price  \
0      1417.391304                        0            326   
1      1552.380952                        0            326   
2      1421.739130                        0            327   
3      1151.724138                        0            334   
4      1080.645161                        0            335   

   rounded_price_to_100multiple  
0                           400  
1                           400  
2                           400  
3                           400  
4                           400

- Eliminar las columnas `rounded_price` y `rounded_price_to_100multiple` utilizando la función `drop`:

In [38]:
diamonds_df = diamonds_df.drop(columns = ['rounded_price', 'rounded_price_to_100multiple'])
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z  \
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43   
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31   
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31   
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63   
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75   

   price_per_carat  price_per_carat_is_high  
0      1417.391304                        0  
1      1552.380952                        0  
2      1421.739130                        0  
3      1151.724138                        0  
4      1080.645161                        0

### Ejercicio 7: Escribir un DataFrame en un archivo

- En este ejercicio, escribiremos un `DataFrame` de diamantes en un archivo `.csv`. Para ello, utilizaremos el siguiente código:

- Cargue la biblioteca `pandas`

In [39]:
import pandas as pd

- Leer los archivos de la `URL` en el `DataFrame` de `pandas`

In [40]:
diamonds_df = pd.read_csv(diamonds_url)
diamonds_df.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

- Escriba el conjunto de datos de los diamantes en un archivo `.csv`:

In [41]:
diamonds_df.to_csv('diamonds_modified.csv')

- Por defecto, la función `to_csv` genera un archivo que incluye las cabeceras de las columnas y los números de las filas. así como los números de fila. Añade un parámetro `index=False` para excluir los números de fila:

In [42]:
diamonds_df.to_csv('diamonds_modified.csv', index = False)

- Ahora que tenemos una idea básica de cómo cargar y manejar los datos en un objeto `DataFrame` de `pandas`, vamos a empezar a hacer algunos gráficos simples a partir de los datos.
- `matplotlib` es una biblioteca de trazado disponible en la mayoría de las distribuciones de `Python` y es la es la base de varios paquetes de ploteo, incluyendo la funcionalidad de ploteo incorporada en `pandas` y `seaborn`. `matplotlib` permite controlar todos los aspectos de una figura y es conocido por ser muy detallado.

### Ejercicio 8: Trazado y análisis de un histograma

- En este ejercicio, crearemos un `histograma` de la frecuencia de los diamantes en el conjunto de datos con sus respectivas especificaciones de `carat` (quilates) en el eje $x$:

- Cargue la biblioteca `pandas`

In [43]:
import pandas as pd
import seaborn as sns

- Leer los archivos de la `URL` en el `DataFrame` de `pandas`

In [ ]:
diamonds_df = pd.read_csv(diamonds_url)
diamonds_df.head()

- Trazar un histograma utilizando el conjunto de datos de diamantes donde el eje $x$ `= carat`. El eje $y$ de este gráfico indica el número de diamantes del conjunto de datos con la especificación de `carat` en el eje $x$.

In [ ]:
diamonds_df.hist(column='carat');

- La función `hist` tiene un parámetro llamado `bins`, que se refiere literalmente al número de intervalos de igual tamaño en los que se dividen los puntos de datos. Por defecto el parámetro `bins` está fijado en 10 en `pandas`. Podemos cambiarlo por un número diferente diferente, si lo deseamos.

In [ ]:
diamonds_df.hist(column = 'carat', bins = 50);

- Ahora, veamos la misma función utilizando `seaborn`. Note que `pandas` establece el parámetro `bins` a un valor por defecto de 10, pero `seaborn` infiere un `bin` de tamaño apropiado basado en la distribución estadística del conjunto de datos.

In [ ]:
import seaborn as sns

In [ ]:
sns.distplot(diamonds_df.carat);

- Por defecto, la función `distplot` también incluye una curva suavizada sobre el histograma, llamada `estimación de la densidad del kernel (KDE)`. Si queremos eliminar el `KDE` y mirar sólo el histograma, podemos utilizar el parámetro `kde=False`.

In [ ]:
sns.distplot(diamonds_df.carat, kde = False);

- Una `transformación logarítmica` ayuda a identificar más tendencias. Por ejemplo, en el siguiente gráfico, el eje $x$ muestra los valores transformados en logaritmos de la variable del precio, y vemos que hay dos picos que indican dos tipos de diamantes: uno con un precio alto y otro con un precio bajo

In [ ]:
import numpy as np

In [ ]:
sns.distplot(np.log(diamonds_df.price));

Qué valores de las características son más frecuentes en el conjunto de datos (en este caso, hay dos picos en torno a 6.8 y 8.5 respectivamente, nótese que `log(price) = valores`, en este caso

### Ejercicio 9: Creación de un gráfico de barras y cálculo de la distribución del precio medio

- En este ejercicio, aprenderemos a crear una tabla utilizando la función `pandas crosstab`. Utilizaremos una tabla para generar un gráfico de barras. A continuación, exploraremos un gráfico de barras generado con la biblioteca `seaborn` y calcularemos la distribución del precio medio. Para ello, vamos a realizar los siguientes pasos

- Cargue la biblioteca **pandas**

In [ ]:
import pandas as pd
import seaborn as sns

- Leer los archivos de la `URL` en el `DataFrame` de `pandas`

In [ ]:
diamonds_df = pd.read_csv(diamonds_url)
diamonds_df.head()

- Imprime los valores únicos de la columna *corte*:

In [ ]:
diamonds_df.cut.unique()

- Imprime los valores únicos de la columna *claridad*:

In [ ]:
diamonds_df.clarity.unique()

- `unique()` devuelve un array. Hay cinco cualidades únicas de *cut* y ocho valores únicos en *clarity*. El número de valores únicos se puede obtener utilizando `nunique()` en `pandas`
- Para obtener los recuentos de diamantes de cada calidad de *cut*, primero creamos una tabla utilizando la función `pandas crosstab()`:

In [ ]:
cut_count_table = pd.crosstab(index = diamonds_df['cut'], columns = 'count')
cut_count_table

- Pase estos recuentos a otra función `pandas`, `plot(kind='bar')`:

In [ ]:
cut_count_table.plot(kind = 'bar');

- Vemos que la mayoría de los diamantes del conjunto de datos son de la calidad de corte `Ideal`, seguidos de `Premium, Very Good, Good y Fair`. Ahora, veamos cómo generar el mismo gráfico utilizando `seaborn`
- Generate the same bar plot using `seaborn`

In [ ]:
sns.catplot("cut", data = diamonds_df, aspect = 1.5, kind = "count", color = "b");

Observe cómo la función `catplot()` no requiere que creemos la tabla de recuento intermedia (utilizando `pd.crosstab()`), y reduce un paso en el proceso de trazado.

- A continuación, se muestra cómo se obtiene la **distribución del precio medio** de las diferentes calidades de `cut` utilizando `seaborn`

In [ ]:
import seaborn as sns
from numpy import median, mean

In [ ]:
sns.set(style = "whitegrid")
ax = sns.barplot(x = "cut", y = "price", data = diamonds_df, estimator = mean)

Aquí, las líneas negras (barras de error) de los rectángulos indican la incertidumbre (o dispersión de los valores) en torno a la estimación de la media. Las funciones mencionadas van mucho más allá de un simple recuento: aplican una función que calcula una medida de tendencia central (por defecto es el valor medio) y muestran, aplicando `bootstrapping`, el intervalo de confianza del 95% para dicha medida. Por defecto, este valor está fijado en un 95% de confianza. ¿Cómo lo cambiamos? Utilizando el parámetro `ci=68`, por ejemplo, lo fijamos en el 68%. También podemos representar la desviación estándar en los precios utilizando `ci=sd`.

- Reordenar las barras del eje $x$ utilizando el orden: 

In [ ]:
ax = sns.barplot(x = "cut", y = "price", data = diamonds_df, estimator = mean, ci = 68, 
                 order=['Ideal','Good','Very Good','Fair','Premium'])

### Ejercicio 10: Creación de gráficos de barras agrupados por una característica específica

- En este ejercicio, utilizaremos el conjunto de datos de los diamantes para generar la distribución de los precios con respecto al color para cada calidad de corte. En el ejercicio 7, vimos la distribución de precios para diamantes de diferentes calidades de talla. Ahora, nos gustaría ver la variación de cada color:

- Cargue la biblioteca **pandas**

In [ ]:
import pandas as pd
import seaborn as sns

- Leer los archivos de la `URL` en el `DataFrame` de `pandas`

In [ ]:
diamonds_df = pd.read_csv(diamonds_url)
diamonds_df.tail()

- Utilice el parámetro `hue` para trazar grupos anidados:

In [ ]:
ax = sns.barplot(x = "cut", y = "price", hue = 'color', data = diamonds_df)

Aquí podemos observar que los `patrones de precios de los diamantes de diferentes colores son similares para cada calidad de talla` `cut`. 

### Ejercicio 11: Cómo ajustar los parámetros de un gráfico de barras agrupadas

- En este ejercicio, modificaremos los parámetros de los gráficos, por ejemplo, `hue`, de un gráfico de barras agrupadas. Veremos cómo colocar las leyendas y las etiquetas de los ejes en los lugares adecuados y también exploraremos la función de rotación

- Cargue la biblioteca **pandas**

In [ ]:
import pandas as pd
import seaborn as sns

- Leer los archivos de la **URL** en el **DataFrame** de **pandas**

In [ ]:
diamonds_df = pd.read_csv(diamonds_url)
diamonds_df.head()

In [ ]:
ax = sns.barplot(x = "cut", y = "price", hue = 'color', data = diamonds_df)

In [ ]:
ax = sns.barplot(x='cut', y='price', hue='color', data=diamonds_df)
ax.legend(loc='upper right', ncol=4);

En la llamada anterior a `ax.legend()`, el parámetro `ncol` denota el número de columnas en las que deben organizarse los valores de la leyenda, y el parámetro `loc` especifica la ubicación de la leyenda y puede tomar cualquiera de los ocho valores `'upper left', 'upper right', 'lower left', 'lower right'`.

- Para modificar las etiquetas de los ejes $𝑥$ y $𝑦$, introduzca el siguiente código

In [ ]:
ax = sns.barplot(x='cut', y='price', hue='color', data=diamonds_df)
ax.legend(loc='upper right', ncol=4)
ax.set_xlabel('Cut', fontdict={'fontsize' : 12})
ax.set_ylabel('Price', fontdict={'fontsize' : 12});

- Del mismo modo, utilice esto para modificar el tamaño de la fuente y la rotación del eje $x$ de la garrapata etiquetas:

In [ ]:
ax = sns.barplot(x='cut', y='price', hue='color', data=diamonds_df)
ax.legend(loc='upper right',ncol=4)
ax.set_xticklabels(ax.get_xticklabels(), fontsize=13, rotation=45)
ax.set_xlabel('Cut', fontdict={'fontsize' : 12})
ax.set_ylabel('Price', fontdict={'fontsize' : 12});

La función de rotación es especialmente útil cuando las etiquetas de ticks son largas y se amontonan en el eje $x$.

### Ejercicio 12: Anotar un gráfico de barras

- En este ejercicio, anotaremos un gráfico de barras, generado con la función **catplot** de **seaborn**, utilizando una nota justo encima del gráfico.

- Cargue la biblioteca **pandas**

In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

- Leer los archivos de la **URL** en el **DataFrame** de **pandas**

In [ ]:
diamonds_df = pd.read_csv(diamonds_url)
diamonds_df.head()

- Generar un gráfico de barras utilizando la función `catplot` de la biblioteca `seaborn`

In [ ]:
ax = sns.catplot("cut", data=diamonds_df, aspect=1.5, kind="count", color="b")

- Anote la columna que pertenece a la categoría `Ideal`:

In [ ]:
ideal_group = diamonds_df.loc[diamonds_df['cut']=='Ideal']
ideal_group.head()

- Encuentre la ubicación de la coordenada $x$ donde debe colocarse la anotación:

In [ ]:
x = ideal_group.index.tolist()[0]

- Encuentre la ubicación de la coordenada $y$ donde debe colocarse la anotación:

In [ ]:
y = len(ideal_group)

- Imprime la ubicación de las coordenadas $x$ e $y$:

In [ ]:
print(x)
print(y)

- Anota la gráfica con una nota:

In [ ]:
sns.catplot("cut", data=diamonds_df, aspect=1.5, kind="count", color="b")
plt.annotate('excellent polish and symmetry ratings;\nreflects almost all the light that enters it', 
             xy=(x,y), xytext=(x+0.3, y+2000), arrowprops=dict(facecolor='red'));

- Ahora, parece que hay muchos parámetros en la función de anotación. La documentación oficial de [Matplotlib](https://matplotlib.org/3.1.0/api/_as_gen/matplotlib.pyplot.annotate.html) cubre todos los detalles

#### Práctica

- Trabajaremos con el conjunto de datos de 120 años de historia olímpica adquirido por Randi Griffin en [Randi-Griffin](https://www.sports-reference.com/) y puesto a disposición en [athlete_events](https://raw.githubusercontent.com/lihkirun/AppliedStatisticMS/main/DataVisualizationRPython/Lectures/Python/PythonDataSets/athlete_events.csv). - Su tarea consiste en identificar los cinco deportes más importantes según el mayor número de medallas otorgadas en el año 2016, y luego realizar el siguiente análisis:

1. Genere un gráfico que indique el número de medallas concedidas en cada uno de los cinco principales deportes en 2016.
2. Trace un gráfico que represente la distribución de la edad de los ganadores de medallas en los cinco principales deportes en 2016.
3. Descubre qué equipos nacionales ganaron el mayor número de medallas en los cinco principales deportes en 2016.
4. Observe la tendencia del peso medio de los atletas masculinos y femeninos ganadores en los cinco principales deportes en 2016

- Pasos principales

1. Descargue el conjunto de datos y formatéelo como un DataFrame de pandas.
2. Filtra el `DataFrame` para incluir solo las filas correspondientes a los ganadores de medallas de 2016.
3. Descubre las medallas concedidas en 2016 en cada deporte.
4. Enumera los cinco deportes más importantes en función del mayor número de medallas concedidas. Filtra el `DataFrame` una vez más para incluir solo los registros de los cinco deportes principales en 2016.
5. Genere un gráfico de barras con los recuentos de registros correspondientes a cada uno de los cinco deportes principales.
6. Generar un histograma para la característica Edad de todos los ganadores de medallas en los cinco deportes principales (2016).
7. Genera un gráfico de barras que indique cuántas medallas ganó el equipo de cada país en los cinco deportes principales en 2016.
8. Genere un gráfico de barras que indique el peso medio de los jugadores, clasificados en función del género, que ganaron en los cinco principales deportes en 2016.

## Visualización estática

- Este capítulo es un manual sobre los diferentes tipos de visualización estática y los contextos en los que son más eficaces. Utilizando **seaborn**, aprenderá a crear una variedad de variedad de gráficos y a seleccionar el tipo correcto de visualización para la representación más adecuada de sus datos.
- En esta sección, estudiaremos el contexto de los gráficos que presentan patrones globales en los datos,
como por ejemplo:
1. Gráficos que muestran la varianza de las características individuales de los datos, como los histogramas
2. Gráficos que muestran cómo varían las diferentes características presentes en los datos entre sí, como los gráficos de dispersión, los gráficos de líneas y los mapas de calor.

- **Scatter plot**

1. Un gráfico de dispersión es un gráfico simple que presenta los valores de dos características en un conjunto de datos.
2. Cada punto de datos se representa mediante un punto con la coordenada $x$ como valor de la primera característica y la coordenada $y$ como valor de la segunda característica. 
3. Un gráfico de dispersión es una gran herramienta para aprender más sobre dos atributos numéricos. 
4. Los gráficos de dispersión pueden ayudar a descubrir las relaciones entre diferentes características de los datos, como el tiempo y las ventas, la ingesta de alimentos y las estadísticas de salud en varios contextos.

### Ejercicio 13: Creación de un gráfico de dispersión estático

- En este ejercicio, generaremos un gráfico de dispersión para examinar la relación entre el *peso* (`weight`) y el *millaje* (`mpg`) de los vehículos del conjunto de datos `mpg`. Para ello, vamos a seguir los siguientes pasos:
- Abra un notebook Jupyter e importe los módulos de Python necesarios:

In [ ]:
import seaborn as sns

- Importe el conjunto de datos de `seaborn`:

In [ ]:
mpg_df = sns.load_dataset("mpg")

- Los datos que utilizamos `mpg` son tomados del repositorio de la UCI [Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/auto+mpg).

In [ ]:
mpg_df.head()

- Generar un gráfico de dispersión utilizando la función `scatterplot()`:

In [ ]:
ax = sns.scatterplot(x="weight", y="mpg", data=mpg_df)

**Observación:** Nótese que el gráfico de dispersión muestra una disminución del *mileage (mpg)* con un aumento del *weight*. Es una visión útil de las relaciones entre las diferentes características del conjunto de datos. Si deseamos calcular la correlación negativa que se puede visualizar usamos la función utilizamos `np.corrcoef(mpg_df["weight"], (mpg_df["mpg"])`

In [ ]:
print("Correlation betewen weight and mpg:", np.corrcoef(mpg_df["weight"], mpg_df["mpg"])[0,1])

### Ejercicio 14: Creación de un gráfico hexagonal estático Binning

- También existe una versión más elegante de los gráficos de dispersión, llamada **hexagonal binning plot (`hexbin plot`)**. En este ejercicio, generaremos un diagrama de dispersión hexagonal para comprender mejor la relación entre el peso (`weight`) y el millaje (mileage (`mpg`)):
- Abra un cuaderno Jupyter e importe los módulos de `Python` necesarios:

In [ ]:
import seaborn as sns

- Importe el conjunto de datos de **seaborn**:

In [ ]:
mpg_df = sns.load_dataset("mpg")

- Trazar un gráfico `hexbin` usando `jointplot` con el tipo establecido en `hex`:

In [ ]:
sns.set(style="ticks")
sns.jointplot(mpg_df.weight, mpg_df.mpg, kind="hex", color="#4CB391");

**Observación:** Como puede observar, el histograma de los ejes superior y derecho representa la varianza de las características representadas por los ejes $x$ e $y$ respectivamente (`mpg` y `weight`, en este ejemplo). Además, es posible que haya notado `en el gráfico de dispersión anterior que los datos se superponían fuertemente en ciertas áreas, ocultando la distribución real de las características`. Los gráficos **Hexbin** son una buena herramienta de visualización utilizada cuando los datos son muy densos.

### Ejercicio 15: Creación de un gráfico de contorno estático

- Otra alternativa a los gráficos de dispersión cuando los puntos de datos están densamente poblados en regiones específicas es un gráfico de contorno. En este ejercicio, crearemos un gráfico de contorno para mostrar la relación entre `weight` y el `mileage` en el conjunto de datos `mpg`. Podremos ver que la relación entre `weight` y `mileage` es más fuerte cuando hay mayor volumen de datos

- Abra un notebook `Jupyter` e importe los módulos de `Python` necesarios:

In [ ]:
import seaborn as sns

- Importe el conjunto de datos de **seaborn**:

In [ ]:
mpg_df = sns.load_dataset("mpg")

- Crea un gráfico de contorno utilizando el método **set_style**

In [ ]:
sns.set_style("white")

- Generar un gráfico de estimación de la densidad del núcleo (KDE). Los dos primeros parámetros son matrices de $X$ e $Y$ coordenadas de los puntos de datos, el parámetro shade se establece en **True** para que los contornos se rellenen con un gradiente de color basado en el número de puntos de datos

In [ ]:
sns.kdeplot(mpg_df.weight, mpg_df.mpg, shade=True);

``Observación:`` En nuestro ejemplo de `weight` frente a `mileage (mpg)`, el diagrama hexagonal y el diagrama de contorno indican que hay una determinada curva a lo largo de la cual la relación negativa entre el peso y el kilometraje es más fuerte, como es evidente por el mayor número de puntos de datos. La relación negativa se vuelve relativamente más débil a medida que nos alejamos de la curva (menos volumen de datos).

### Ejercicio 16: Creación de un gráfico de líneas estáticas

- Los gráficos de líneas representan la información como una serie de puntos de datos conectados por segmentos de líneas rectas. Son útiles para indicar la relación entre una característica numérica discreta (en el eje $x$), como `model_year`, y una característica numérica continua (en el eje $y$), como *mpg* del conjunto de datos `mpg`. En este ejercicio, crearemos un gráfico de dispersión para un par de características diferentes, `model_year` y `mpg`.
- Abra un cuaderno `Jupyter` e importe los módulos de `Python` necesarios:

In [ ]:
import seaborn as sns

- Importe el conjunto de datos de **seaborn**:

In [ ]:
mpg_df = sns.load_dataset("mpg")

- Slecciones el estilo para la figura con `set_style`

In [ ]:
sns.set_style("white")

- Crear un gráfico de dispersión bidimensional

In [ ]:
ax1 = sns.scatterplot(x="model_year", y="mpg", data=mpg_df)

**Observación**: En este ejemplo, vemos que la característica `model_year` sólo toma valores discretos entre 70 y 82. Ahora, cuando tenemos una característica numérica discreta como esta (`model_year`), dibujar un gráfico de líneas uniendo los puntos de datos es una buena idea. Podemos dibujar un simple gráfico de líneas que muestre la relación entre model_year y mpg con el siguiente código.

- Dibuja un gráfico lineal simple para mostrar la relación entre `model_year` y `mileage`

In [ ]:
ax = sns.lineplot(x="model_year", y="mpg", data=mpg_df)

**Observación:** Como podemos ver, los puntos conectados por la línea sólida representan la media de la característica del eje $y$ en la coordenada $x$ correspondiente. El área sombreada alrededor de la línea muestra el intervalo de confianza para la característica del eje $y$ (por defecto, `seaborn` establece el intervalo de confianza del 95%). El parámetro `ci` puede utilizarse para cambiar a un intervalo de confianza diferente.

- Cambiar el intervalo de confianza a 68

In [ ]:
sns.lineplot(x="model_year", y="mpg", data=mpg_df, ci=68);

**Observación:** Como podemos ver en el gráfico anterior, el intervalo de confianza del 68% se traduce en un rango de valores de características en el que están presentes el 68% de los puntos de datos. Los gráficos de líneas son excelentes técnicas de visualización para escenarios en los que tenemos datos que cambian con el tiempo, eje $x$, podría representar la fecha o el tiempo, y el gráfico ayudaría a visualizar cómo un valor varía a lo largo de ese periodo

### Ejercicio 17: Presentación de datos a través del tiempo con múltiples gráficos de líneas

- En este ejemplo, veremos cómo presentar los datos a través del tiempo con múltiples gráficos de líneas. En este ejemplo utilizamos el conjunto de datos de vuelos de `seaborn`
- Abra un cuaderno Jupyter e importe los módulos de Python necesarios

In [ ]:
import seaborn as sns

- Importe el conjunto de datos de `seaborn`:

In [ ]:
flights_df = sns.load_dataset("flights")
print(flights_df.head())

- Supongamos que quiere observar cómo varía el número de pasajeros entre meses en diferentes años. ¿Cómo mostraría esta información? Una opción es dibujar varios gráficos de líneas en una sola figura.
- Crear varias gráficas para los meses de diciembre y enero

In [ ]:
fig,ax=plt.subplots()
ax = sns.lineplot(x="year", y="passengers", data=flights_df[flights_df['month']=='Jan'], color='green')
ax = sns.lineplot(x="year", y="passengers", data=flights_df[flights_df['month']=='Feb'], color='red')
ax = sns.lineplot(x="year", y="passengers", data=flights_df[flights_df['month']=='Mar'], color='blue')
ax = sns.lineplot(x="year", y="passengers", data=flights_df[flights_df['month']=='Apr'], color='cyan')
ax = sns.lineplot(x="year", y="passengers", data=flights_df[flights_df['month']=='May'], color='pink')
ax = sns.lineplot(x="year", y="passengers", data=flights_df[flights_df['month']=='Jun'], color='black')
ax = sns.lineplot(x="year", y="passengers", data=flights_df[flights_df['month']=='Jul'], color='grey')
ax = sns.lineplot(x="year", y="passengers", data=flights_df[flights_df['month']=='Aug'], color='yellow')
ax = sns.lineplot(x="year", y="passengers", data=flights_df[flights_df['month']=='Sep'], color='turquoise')
ax = sns.lineplot(x="year", y="passengers", data=flights_df[flights_df['month']=='Oct'], color='orange')
ax = sns.lineplot(x="year", y="passengers", data=flights_df[flights_df['month']=='Nov'], color='darkgreen')
ax = sns.lineplot(x="year", y="passengers", data=flights_df[flights_df['month']=='Dec'], color='darkred')

**Observación** Con este ejemplo de 12 gráficos de líneas, podemos ver cómo una figura con demasiados gráficos de líneas rápidamente comienza a confundirse. Por lo tanto, para ciertos escenarios, los gráficos de líneas no son no son atractivos ni útiles. Entonces, ¿cuál es la alternativa para nuestro caso de uso? Lo veremos en el siguiente ejercicio

### Ejercicio 18: Creación y exploración de un mapa de calor estático

- Un `heatmap` es una representación visual de una característica numérica continua específica en función de otras dos características discretas (ya sea una categórica o una numérica discreta) en el conjunto de datos. En este ejercicio, exploraremos y crearemos un mapa de calor. Utilizaremos el conjunto de datos de vuelos de la biblioteca `seaborn` para generar un `mapa de calor` que represente el número de pasajeros por mes en los años 1949-1960
- Abra un cuaderno `Jupyter` e importe los módulos de `Python` necesarios:

In [ ]:
import seaborn as sns

- Importe el conjunto de datos de `seaborn`:

In [ ]:
flights_df = sns.load_dataset("flights")
print(flights_df.head())

- Ahora tenemos que hacer pivotar el conjunto de datos sobre las variables requeridas utilizando la función `pivot()` antes de generar el mapa de calor. La función `pivot` toma primero como argumentos la característica que se mostrará en filas, luego la que se muestra en columnas y, por último, la característica cuya variación nos interesa observar. Utiliza los valores únicos de los índices/columnas especificados para formar los ejes del `DataFrame` resultante

In [ ]:
df_pivoted = flights_df.pivot("month", "year", "passengers")
ax = sns.heatmap(df_pivoted)

**Observación:** Aquí podemos observar que el número total de vuelos anuales aumentó de forma constante desde 1949 a 1960. Además, los meses de julio y agosto parecen tener el mayor número de vuelos (en comparación con el resto de los meses) en todos los años de observación.

- Utilice la opción `clustermap` para agrupar filas y columnas:

In [ ]:
ax = sns.clustermap(df_pivoted, col_cluster=False, row_cluster=True)

**Observación:** Nótese que el orden de los meses se ha reordenado en los gráficos, pero algunos meses (por ejemplo, julio y agosto) se han mantenido juntos debido a sus **tendencias similares**. Tanto en julio como en agosto, el número de vuelos aumentó de forma relativamente más drástica en los últimos años hasta 1960. ¿Cómo se calcula la similitud entre filas y columnas? La respuesta es que depende de la métrica de distancia

- Establecer la métrica como euclidiana

In [ ]:
ax = sns.clustermap(df_pivoted, col_cluster=False)

- Cambiar la métrica a correlación

In [ ]:
ax = sns.clustermap(df_pivoted, row_cluster=False, metric='correlation')

**Observación:** Al leer sobre la métrica de la distancia, aprendemos que define la distancia entre dos filas/columnas. Sin embargo, si nos fijamos bien, vemos que el mapa de calor también agrupa no sólo filas o columnas individuales, sino también grupos de filas o columnas individuales. Aquí es donde entra en juego la vinculación.

### Ejercicio 19: Creación de vínculos en mapas térmicos estáticos

1. Si definimos la distancia entre dos clusters como la **distancia entre los dos puntos de los clusters más cercanos** entre sí, la regla se denomina enlace único (**single linkage**).
2. Si la regla es definir la distancia entre dos clusters como la **distancia entre los puntos más alejados entre sí**, se denomina vinculación completa (**complete linkage**).
3. Si la regla es definir la distancia como **la media de todos los posibles pares de filas en los dos clústeres**, se denomina vinculación media (**average linkage**).

- En este ejercicio, generaremos un mapa de calor y comprenderemos el concepto de enlace único, completo y promedio en los mapas de calor utilizando el conjunto de datos `flights`.
- Abra un cuaderno Jupyter e importe los módulos de Python necesarios:

In [ ]:
import seaborn as sns

- Importe el conjunto de datos de `seaborn`:

In [ ]:
flights_df = sns.load_dataset("flights")
print(flights_df.head())

- Ahora necesitamos pivotar el conjunto de datos en las variables requeridas utilizando la función `pivot()` antes de generar el mapa de calor:

In [ ]:
df_pivoted = flights_df.pivot("month", "year", "passengers")
ax = sns.heatmap(df_pivoted)

- Enlaza los mapas de calor utilizando el código que sigue

In [ ]:
ax = sns.clustermap(df_pivoted, col_cluster=False, metric='correlation', method='average')

In [ ]:
ax = sns.clustermap(df_pivoted, row_cluster=False, metric='correlation', method='complete')

In [ ]:
ax = sns.clustermap(df_pivoted, row_cluster=False, metric='correlation', method='single')

**Observación** Los mapas de calor también son una buena forma de visualizar lo que ocurre en un espacio 2D. Por ejemplo, pueden utilizarse para mostrar dónde hay más acción en el campo en un partido de fútbol. Del mismo modo, en un sitio web, **los mapas de calor se pueden utilizar para mostrar las áreas que son más más frecuentadas por los usuarios**.

#### Creación de gráficos para representar estadísticas

- Cuando los conjuntos de datos son enormes, a veces resulta útil observar las estadísticas de resumen de una serie de características diferentes y hacerse una idea preliminar del conjunto de datos. Por ejemplo, las estadísticas de resumen de cualquier característica numérica incluyen **medidas de tendencia central, como la media, y medidas de dispersión, como la desviación estándar**. Los histogramas muestran la distribución de una característica dada en los datos, podemos hacer un gráfico un poco más informativo mostrando algunas estadísticas de resumen en el mismo gráfico.

#### Ejemplo 1: Histograma revisado

- Importar los módulos de Python necesarios; cargar el conjunto de datos; elegir el número de intervalos y si la estimación de la densidad del kernel debe mostrarse o no; usar el color rojo para mostrar media utilizando una línea recta en el eje $x$ (paralela al eje $y$); definir la ubicación de la leyenda:

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [ ]:
mpg_df = sns.load_dataset("mpg")
ax = sns.distplot(mpg_df.weight, bins=50, kde=False)
plt.axvline(x=np.mean(mpg_df.weight), color='red', label='mean')
plt.axvline(x=np.median(mpg_df.weight), color='orange', label='median')
plt.legend(loc='upper right');

Este histograma muestra la distribución de la característica de peso junto con la media y la mediana. Observe que **la media no es igual a la mediana, lo que significa que la característica no está distribuida normalmente**.

### Ejercicio 20: Creación y exploración de un gráfico de caja

Los gráficos de caja son una forma excelente de examinar la relación entre las estadísticas de resumen de una característica numérica en relación con otras características categóricas. En este ejercicio crearemos un diagrama para establecer la relación entre `model_year` and `mileage` using the `mpg` dataset. We'll analyze manufacturing efficiency and the mileage of vehicles over a period of years. To do so, let's go through the following steps

- Importe la libreria `seaborn`

In [ ]:
import seaborn as sns

- Cargue el `dataset`

In [ ]:
mpg_df = sns.load_dataset("mpg")
mpg_df.head()

- Crear un diagrama de cajas

In [ ]:
sns.boxplot(x='model_year', y='mpg', data=mpg_df);

Como podemos ver, los límites de la caja indican el rango intercuartil, el límite superior marca el cuartil del 25% y el límite inferior el cuartil del 75%. La línea horizontal dentro de la caja indica la mediana. Cualquier punto aislado fuera de los bigotes (las barras en forma de $T$ por encima y por debajo de la caja) marca los valores atípicos, mientras que los propios bigotes muestran los valores mínimos y máximos que no son valores atípicos. Aparentemente, el millaje mejoró sustancialmente en los 80 en comparación con los 70.

- Añadamos otra característica a nuestro `mpg` DataFrame que denote si el coche fue fabricado en los 70 o en los 80. Modifica el DataFrame `mpg` creando una nueva característica, `model_decade`

In [ ]:
import numpy as np
mpg_df['model_decade'] = np.floor(mpg_df.model_year/10)*10
mpg_df['model_decade'] = mpg_df['model_decade'].astype(int)
mpg_df.tail()

- Ahora, volvamos a dibujar nuestro gráfico de caja para ver la distribución de los millajes en las dos décadas

In [ ]:
sns.boxplot(x='model_decade', y='mpg', data=mpg_df)

También podemos añadir otra característica, por ejemplo, la región de origen, y ver cómo afecta a la relación entre el kilometraje y el año de fabricación, las dos características que hemos estado considerando hasta ahora

- Utiliza el parámetro `hue` para agrupar por `origin`

In [ ]:
sns.boxplot(x='model_decade', y='mpg', data=mpg_df, hue='origin');

Como podemos ver, según el conjunto de datos de mpg, en los años 70 y principios de los 80, Europa y Japón produjeron coches con mejor millaje que los Estados Unidos.

### Ejercicio 21: Creación de un gráfico de violín

- ¿Y si pudiéramos obtener una pista sobre la distribución completa de una característica numérica específica agrupada por otras características categóricas? La tipo de técnica de visualización adecuada en este caso es un gráfico de violín. Un gráfico de violín es similar a un grafico de caja pero incluye más detalles sobre las variaciones de los datos. En este ejercicio, utilizaremos el conjunto de datos **mpg** y generaremos un gráfico de violín que represente la variación detallada del **millaje (mpg)** en función de **model_decade** y la región de origen

- Importe la libreria **seaborn**

In [ ]:
import seaborn as sns

- Cargue el **dataset**

In [ ]:
mpg_df = sns.load_dataset("mpg")

- Generar el gráfico de violín utilizando la función **violinplot** en **seaborn**

In [ ]:
import numpy as np

mpg_df['model_decade'] = np.floor(mpg_df.model_year/10)*10
mpg_df['model_decade'] = mpg_df['model_decade'].astype(int)
sns.violinplot(x='model_decade', y='mpg', data=mpg_df, hue='origin')

- Podemos ver aquí que, durante los años 70, mientras que la mayoría de los vehículos de EE.UU. tenían un millaje medio de 19 mpg, los vehículos de Japón y Europa tenían millajes medios de alrededor de 27 y 25 mpg. Mientras que el kilometraje de los vehículos en Europa y Japón aumentó entre 7 y 8 puntos en los años 80, el kilometraje medio de los vehículos en EE.UU. seguía siendo similar al de los vehículos en Japón y Europa en la década anterior.

#### Práctica

**Estadísticas:** Seguiremos trabajando con el conjunto de datos de 120 años de historia olímpica adquirido por Randi Griffin en [Randi Griffin](https://www.sports-reference.com/)

Como especialista en visualización, su tarea consiste en crear dos parcelas para los ganadores de medallas de 2016 de cinco deportes: atletismo, natación, remo, fútbol y hockey

- Crea un gráfico utilizando una técnica de visualización adecuada que presente de la mejor manera posible el patrón global de las características de **height** y **weight** de los ganadores de medallas de 2016 de los cinco deportes.
- Crea un gráfico utilizando una técnica de visualización adecuada que presente de la mejor manera posible la estadística de resumen para la altura y el peso de los jugadores que ganaron cada tipo de medalla (oro/plata/bronce) en los datos.

Utilizar su creatividad y sus habilidades para sacar conclusiones importantes de los datos

**Pasos importantes**

- Descargue el conjunto de datos y formatéelo como un **pandas** DataFrame
- Filtrar el DataFrame para incluir únicamente las filas correspondientes a los ganadores de medallas de 2016 en los deportes mencionados en la descripción de la actividad
- Observe las características del conjunto de datos y anote su tipo de datos: ¿son categóricos o numéricos?
- Evaluar cuál sería la visualización adecuada para que un patrón global represente las características de **height** y **weight**
- Evaluar cuál sería la visualización adecuada para representar las estadísticas resumidas de las características de **height** y **weight** en función de las medallas, separadas además por género de los atletas.